casper/src/main/resources/MultiSigRevVault.rho

```diff
@@ -0,0 +1,215 @@
+ 4.  | 1,         | secp256k1          | pk = 04fe2eb1e0e7462b1a8f64600389e1e76727f8b2d38804eaa4b48f7a7d6715130fc24d3c4dac2d8bdc19e0b49879dbaf07c30773cd9740a9d14a092ef76339207a
+ 5.  | 4, 2,      | genIds             | uname = Unforgeable(0x022e67c5237b24bcc1971f00556aa180f8d864c67e23989c9d550f8c9265cb53)
+ 6.  | 3, 5,      | registry           | value = (9223372036854775807, bundle+ {   Unforgeable(0x022e67c5237b24bcc1971f00556aa180f8d864c67e23989c9d550f8c9265cb53) })
+ 7.  | 6,         | protobuf           | toSign = 2a40aa013d0a0d2a0b10feffffffffffffffff010a2c5a2a0a263a240a220a20022e67c5237b24bcc1971f00556aa180f8d864c67e23989c9d550f8c9265cb531001
+ 8.  | 7, 1,      | secp256k1          | sig = 30450221008539caafe08b2a27cfcabdcf9c7c345a1c1a926330a1a320ca614cf9240f6c77022024ad8465597c8a673431007b8e33317bd49c9c68dacb41eef6a4316d933909c4
+ 9.  | 4,         | registry           | uri = rho:id:b9s6j3xeobgset4ndn64hje64grfcj7a43eekb3fh43yso5ujiecfn
+ ----+------------+--------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------
+ */
+
+new MultiSigRevVault,
+    rs(`rho:registry:insertSigned:secp256k1`), uriOut,
+    rl(`rho:registry:lookup`), 
+    _fromPublicKey,
+    _multiSigRevVault,
+    RevVaultCh,
+    AuthKeyCh,
+    ListOpsCh
+in {
+  rl!(`rho:lang:listOps`, *ListOpsCh) |
+  rl!(`rho:rchain:authKey`, *AuthKeyCh) |
+  rl!(`rho:rchain:revVault`, *RevVaultCh) |
+  for(@(_, ListOps) <- ListOpsCh;
+      @(_, AuthKey) <- AuthKeyCh;
+      @(_, RevVault) <- RevVaultCh) {
+
+    contract MultiSigRevVault(@"foo", ret) = { ret!("foo") } |
+
+    contract MultiSigRevVault(@"deployerAuthKey", deployerId, ret) = {
+      new RevAddress(`rho:rev:address`), 
+          DeployerIdOps(`rho:rchain:deployerId:ops`), 
+          revAddrCh, 
+          deployerPubKeyBytesCh in {
+        DeployerIdOps!("pubKeyBytes", *deployerId, *deployerPubKeyBytesCh) |
+        for (@deployerPubKeyBytes <- deployerPubKeyBytesCh) {
+          _fromPublicKey!(deployerPubKeyBytes, *ret)
+        }
+      }
+    } |
+    
+    contract MultiSigRevVault(@"unforgeableAuthKey", unf, ret) = {
+      new RevAddress(`rho:rev:address`), revAddrCh in {
+        RevAddress!("fromUnforgeable", *unf, *revAddrCh) |
+        for (@unfRevAddress <- revAddrCh) {
+          @AuthKey!("make", (*_multiSigRevVault, unfRevAddress), *ret)
+        }
+      }
+    } |
+
+    contract _fromPublicKey(@publicKey, ret) = {
+      new RevAddress(`rho:rev:address`), revAddrCh in {
+        RevAddress!("fromPublicKey", publicKey, *revAddrCh)|
+        for (@revAddress <- revAddrCh) {
+          @AuthKey!("make", (*_multiSigRevVault, revAddress), *ret)
+        }
+      }
+    } |
+    
+    // MultiSig
+    // Returns (false, errorMsg) or (true, (multiSigVault, revAddress, revVault))
+    contract MultiSigRevVault(@"create", @publicKeys, @authKeys, @quorumSize, ret) = {
```

Comment:
So you're worried that if there are two contracts that want to form a multisig, one of them has to trust the other?  That's reasonable.  At the moment that threat can be mitigated to some extent by doing a code analysis, but that's notoriously unreliable without formal verification.

If I were designing it, I would have used an unsealer (the ocaps version of a public key) instead of an authKey.

Reply:
Yes, that's what I'm worried about. I don't really see a point in creating a multisig with `authKeys` when the caller of `Multisig!create` has access to all `authKeys`.

We could provide a `create` method that takes `size`, `publicKeys` and `quorumSize` and returns a contract that then can be used to safely collect `authKeys`.

Reply:
It doesn't necessarily have access to all authkeys.  If public keys are used, there's no way to derive the authkeys from them.  Authkeys derived from unforgeable names are only for use by contracts.

An alternative would be for contracts to provide a name of the form `bundle+{P}`, where only the contract can read from it, and the create call creates an unforgeable name, derives an authkey from it, and sends that authkey on the supplied channel.  That would eliminate the need to trust the caller of `create`, but it deviates somewhat from the design doc.

Reply:
> Authkeys derived from unforgeable names are only for use by contracts.

Not sure what you mean here? If two contracts derive an `authKey` from an unforgeable name, there still is the problem of bringing the authKeys together in a safe way to then call `Multisig!create`, right?

I agree that everything is fine if only `publicKeys` are provided.

Reply:
I can alter the design to use sealer/unsealer pairs instead of public/private keys.  Instead of providing the `create` call with an authkey from an unforgeable name, a contract can provide an unsealer.  The unsealer is the ocap version of a signature verification key.

When confirming a transaction, the contract would provide a sealed tuple `(multisig, nonce)` instead of the authkey.  Since that tuple is only ever used once, there's no possibility of a replay attack.

Reply:
I updated the code to use unsealers for contracts instead of authKeys.  When calling transfer or confirm, the contract needs to seal all the parameters for the call just like a person signs the content when constructing a deploy.  Instead of exposing an `unforgeableAuthKey` method, it now exposes a `makeSealerUnsealer` method.

Reply:
Sorry @pyrocto, I had to switch over to reviewing cost accounting. Once I'm done there I'll review your changes.